### ISSUE TITLE
CMake build script should strive to be package-manager-agnostic

### ISSUE THREAD

**Issue by mbeutel** [KazDragon/telnetpp#214]
The CMakeLists.txt build script currently uses Conan to get dependencies. Although this kind of usage is encouraged by Conan, it unnecessarily couples the build script to Conan itself. Ideally, the build script would use standard CMake means of referring to packages (`find_package()`), and it would be up for the user to decide whether to use Conan, Vcpkg, or another package manager.

telnetpp is also available via Vcpkg. This is currently achieved by patching out the Conan-specific parts and replacing them with CMake default behavior:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 09b4f9b..6f72c41 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -22,8 +22,8 @@ message("Building Telnet++ with zlib: ${TELNETPP_WITH_ZLIB}")
 message("Building Telnet++ with code coverage: ${TELNETPP_COVERAGE}")
 message("Building Telnet++ with sanitizers: ${TELNETPP_SANITIZE}")
 
-include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
-conan_basic_setup(TARGETS NO_OUTPUT_DIRS)
+find_package(gsl-lite REQUIRED)
+find_package(Boost REQUIRED COMPONENTS container)
 
 # The required C++ Standard for Telnet++ is C++14.
 set(CMAKE_CXX_STANDARD 14)
@@ -154,18 +154,24 @@ add_library(telnetpp
     ${TELNETPP_PRIVATE_INCLUDE_FILES}
 )
 
 target_link_libraries(telnetpp
     PUBLIC
-        CONAN_PKG::gsl-lite
-        CONAN_PKG::boost_container
-        CONAN_PKG::boost_signals2
-        CONAN_PKG::boost_variant
+        gsl::gsl-lite
+        Boost::boost
+        Boost::container
 )
 
 if (TELNETPP_WITH_ZLIB)
+    find_package(ZLIB REQUIRED)
     target_link_libraries(telnetpp
         PRIVATE
-            CONAN_PKG::zlib
+            ZLIB::ZLIB
     )
 endif()
 
@@ -179,8 +185,7 @@ set_target_properties(telnetpp
 target_include_directories(telnetpp
     PUBLIC
         $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
        $<INSTALL_INTERFACE:include/telnetpp-${TELNETPP_VERSION}>
-        "${Boost_INCLUDE_DIRS}"
 )
```

(patch based on https://github.com/mbeutel/vcpkg/blob/feature/gsl-lite-0.36/ports/telnetpp/fix-build-error.patch)

Like Vcpkg, Conan can also be used with a toolchain so as to avoid Conan-specific details in the CMakeLists.txt build scripts:
https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_paths_generator.html#included-as-a-toolchain

**Comment by KazDragon** [comment]
For some reason, I've only just noticed this ticket, I apologise for not paying it more attention sooner.

I saw from your other PR that this is in Vcpkg, which surprised the heck out of me.  I'm curious as to whether it's actually being used, or is just there for completeness.

In any case, I've also been investigating the cmake_paths_generator independently, and haven't got it working entirely well.  Allowing Telnet++ to be used both with and independently of Conan is definitely a worthy goal, though.

**Comment by mbeutel** [comment]
> I saw from your other PR that this is in Vcpkg, which surprised the heck out of me. I'm curious as to whether it's actually being used, or is just there for completeness.

The initial port was contributed in https://github.com/microsoft/vcpkg/commit/6e6c2c99485902c593f3b7a21ef265d2db03c1f2 by someone who isn't otherwise working on Vcpkg, so I assume it was added to satisfy a demand, not just for completeness. I got involved with your library because I had taken up maintenance of [gsl-lite](https://github.com/gsl-lite/gsl-lite), and recent changes to Vcpkg's gsl-lite portfile necessitated an adjustment of Vcpkg's patches for Telnet++.


> In any case, I've also been investigating the cmake_paths_generator independently, and haven't got it working entirely well.

Yeah, Conan's ["transparent CMake integration"](https://blog.conan.io/2018/06/11/Transparent-CMake-Integration.html) isn't really transparent; as far as I understand, it insists on generating find modules for all packages it manages rather than letting config modules do the work, which can lead to incompatible semantics (e.g. having to use `find_package(libcurl)` rather than `find_package(CURL)`).

I'm not really sure what to do here. Personally I'd prefer to avoid any hard dependency on Conan, or any other package manager, in library build scripts. (The [blog post I referred to](https://blog.conan.io/2018/06/11/Transparent-CMake-Integration.html) suggests to "patch your CMakeLists.txt from the recipe" to apply Conan specifics to a Conan-agnostic build script.)

**Comment by KazDragon** [comment]
That's really good information.  Thanks.  I guess I have a little more exploration to do to make this whole things more useful.

**Comment by KazDragon** [comment]
I have spent quite some time considering and experimenting on how to make this work.  In particular, how I could write configure the project in such a way that it could be built either with local dependencies or with Conan dependencies.  Something blocks me either way.  

The transparent CMake integration doesn't work for me: it gets stuck in what appears to be an infinite loop during the Boost dependency lookup (or if it isn't infinite, it's just way, way, way too long for the case that everything's right there in place already).

I have also considered writing the CMake script so that it could be built with/without Conan integration and lookup packages as appropriate.  However, this turns into an if/else hell, which isn't helped by CMake's inflexibility when it comes to declaring aliases.

In short, unless something in the tooling to explicitly support this process, I'm not going to be doing anything with it.

I really do thank you for taking the time to bring that to my attention.

**Comment by KazDragon** [comment]
I am reopening this, because with a little help from the guys on the C++ slack Conan channel, I've finally sussed out how it's done.  In an upcoming PR, the repo will have suddenly forgotten about Conan and be looking for system dependencies, but also be updated in my Conan repo.
